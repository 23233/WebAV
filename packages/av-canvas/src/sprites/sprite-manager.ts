import { VisibleSprite, EventTool } from '@webav/av-cliper';

export enum ESpriteManagerEvt {
  ActiveSpriteChange = 'activeSpriteChange',
  AddSprite = 'addSprite',
}

export class SpriteManager {
  #sprites: VisibleSprite[] = [];

  #activeSprite: VisibleSprite | null = null;

  #evtTool = new EventTool<{
    [ESpriteManagerEvt.AddSprite]: (s: VisibleSprite) => void;
    [ESpriteManagerEvt.ActiveSpriteChange]: (s: VisibleSprite | null) => void;
  }>();

  audioCtx = new AudioContext();

  audioMSDest = this.audioCtx.createMediaStreamDestination();

  on = this.#evtTool.on;

  get activeSprite(): VisibleSprite | null {
    return this.#activeSprite;
  }
  set activeSprite(s: VisibleSprite | null) {
    if (s === this.#activeSprite) return;
    this.#activeSprite = s;
    this.#evtTool.emit(ESpriteManagerEvt.ActiveSpriteChange, s);
  }

  constructor() {
    this.#bgAudio();
  }

  async addSprite(vs: VisibleSprite): Promise<void> {
    await vs.initReady;
    this.#sprites.push(vs);
    this.#sprites = this.#sprites.sort((a, b) => a.zIndex - b.zIndex);
    // todo: remove audioNode
    vs.audioNode?.connect(this.audioMSDest);

    this.#evtTool.emit(ESpriteManagerEvt.AddSprite, vs);
  }

  /**
   * Sort in ascending order by sprite.zIndex
   */
  sortSprite() {
    // todo: auto sort
    // this.#sprites = this.#sprites.sort((a, b) => a.zIndex - b.zIndex);
  }

  removeSprite(spr: VisibleSprite): void {
    this.#sprites = this.#sprites.filter((s) => s !== spr);
    spr.destroy();
  }

  getSprites(): VisibleSprite[] {
    return [...this.#sprites];
  }

  destroy(): void {
    this.#evtTool.destroy();
    this.#sprites.forEach((s) => s.destroy());
    this.#sprites = [];
    this.audioMSDest.disconnect();
    this.audioCtx.close().catch(console.error);
  }

  // 添加背景音，如果没有音频，录制的webm不正常
  #bgAudio(): void {
    const oscillator = this.audioCtx.createOscillator();
    const wave = this.audioCtx.createPeriodicWave(
      new Float32Array([0, 0]),
      new Float32Array([0, 0]),
      { disableNormalization: true },
    );
    oscillator.setPeriodicWave(wave);
    oscillator.connect(this.audioMSDest);
    oscillator.start();
  }
}
