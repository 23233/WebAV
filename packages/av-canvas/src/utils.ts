export function createEl (tagName: string): HTMLElement {
  return document.createElement(tagName)
}

// todo: 合并，依赖 cliper
export function renderTxt2Img (txt: string, cssText: string): HTMLImageElement {
  const div = createEl('div')
  div.style.cssText = `${cssText} visibility: hidden; position: fixed;`
  div.textContent = txt
  document.body.appendChild(div)

  const { width, height } = div.getBoundingClientRect()
  // 计算出 rect，立即从dom移除
  div.remove()
  div.style.visibility = 'visible'

  const img = new Image()
  img.width = width
  img.height = height
  const svgStr = `
    <svg xmlns="http://www.w3.org/2000/svg" width="${width}" height="${height}">
    <foreignObject width="100%" height="100%">
        <div xmlns="http://www.w3.org/1999/xhtml">${div.outerHTML}</div>
    </foreignObject>
    </svg>
  `.replace(/\n/g, '').replace(/\t/g, '').replace(/#/g, '%23')

  img.src = `data:image/svg+xml;charset=utf-8,${svgStr}`
  return img
}
