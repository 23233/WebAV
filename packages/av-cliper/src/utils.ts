import mp4box, { MP4File, TrakBoxParser } from 'mp4box'

export function convertFile2Stream (
  file: MP4File,
  timeSlice: number,
  onCancel: () => void
): {
    stream: ReadableStream<ArrayBuffer>
    stop: () => void
  } {
  let timerId = 0

  let sendedBoxIdx = 0
  const boxes = file.boxes
  const deltaBuf = (): ArrayBuffer | null => {
    if (sendedBoxIdx >= boxes.length) return null
    const ds = new mp4box.DataStream()
    ds.endianness = mp4box.DataStream.BIG_ENDIAN
    for (let i = sendedBoxIdx; i < boxes.length; i++) {
      boxes[i].write(ds)
    }
    sendedBoxIdx = boxes.length
    return ds.buffer
  }

  let stoped = false
  let exit: (() => void) | null = null
  const stream = new ReadableStream({
    start (ctrl) {
      timerId = self.setInterval(() => {
        const buf = deltaBuf()
        console.log(1222, buf)
        if (buf != null) ctrl.enqueue(buf)
      }, timeSlice)

      exit = () => {
        clearInterval(timerId)
        file.flush()
        const buf = deltaBuf()
        if (buf != null) ctrl.enqueue(buf)
        ctrl.close()
      }

      // 安全起见，检测如果start触发时已经 stoped
      if (stoped) exit()
    },
    cancel () {
      clearInterval(timerId)
      onCancel()
    }
  })

  return {
    stream,
    stop: () => {
      stoped = true
      exit?.()
    }
  }
}

// track is H.264 or H.265.
export function parseVideoCodecDesc (track: TrakBoxParser): Uint8Array {
  for (const entry of track.mdia.minf.stbl.stsd.entries) {
    if ('avcC' in entry || 'hvcC ' in entry) {
      const stream = new mp4box.DataStream(
        undefined,
        0,
        mp4box.DataStream.BIG_ENDIAN
      )
      // @ts-expect-error
      const box = 'avcC' in entry ? entry.avcC : entry.hvcC
      box.write(stream)
      return new Uint8Array(stream.buffer, 8) // Remove the box header.
    }
  }
  throw Error('avcC or hvcC not found')
}

export async function sleep (time: number): Promise<void> {
  return await new Promise((resolve) => {
    setTimeout(resolve, time)
  })
}
