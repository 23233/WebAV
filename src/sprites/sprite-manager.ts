import { EventTool } from '../event-tool'
import { BaseSprite } from './base-sprite'

export class SpriteManager {
  #sprites: BaseSprite[] = []

  #activeSprite: BaseSprite | null = null

  #evtTool = new EventTool<{
    add: (s: BaseSprite) => void
  }>()

  on = this.#evtTool.on

  get activeSprite (): BaseSprite | null { return this.#activeSprite }
  set activeSprite (s: BaseSprite | null) {
    if (s === this.#activeSprite) return
    if (s == null) {
      if (this.#activeSprite != null) this.#activeSprite.actived = false
    } else {
      s.actived = true
    }
    this.#activeSprite = s
  }

  addSprite<S extends BaseSprite>(s: S): void {
    this.#sprites.push(s)
    this.#sprites = this.#sprites.sort((a, b) => a.zIndex - b.zIndex)
    this.#evtTool.emit('add', s)
    // todo: 动态适配canvas宽高
  }

  getSprites (): BaseSprite[] {
    return [...this.#sprites]
  }
}
