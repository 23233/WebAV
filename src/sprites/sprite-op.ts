import { BaseSprite, SpriteManager } from '.'
import { Rect, TCtrlKey } from './rect'

/**
 * 让canvas中的sprite可以被拖拽移动
 */
export function draggabelSprite (
  cvsEl: HTMLCanvasElement,
  sprMng: SpriteManager
): () => void {
  const cvsRatio = {
    w: cvsEl.clientWidth / cvsEl.width,
    h: cvsEl.clientHeight / cvsEl.height
  }

  // 排在后面的层级更高
  let sprList = sprMng.getSprites().reverse()
  const offAddSpr = sprMng.on('add', () => {
    sprList = sprMng.getSprites().reverse()
  })
  let offScale = (): void => {}

  let startX = 0
  let startY = 0
  let hitSpr: BaseSprite | null = null
  let rect: Rect | null = null
  let limit: Record<'xl' | 'xr' | 'yt' | 'yb', number> | null = null

  // 按下cvs，寻找选中的 sprite，监听移动事件
  const onCvsMouseDown = (evt: MouseEvent): void => {
    // 鼠标左键才能拖拽移动
    if (evt.button !== 0) return
    const { offsetX, offsetY, clientX, clientY } = evt
    hitSpr = sprList.find(s => s.rect.checkHit(
      offsetX / cvsRatio.w,
      offsetY / cvsRatio.h
    )) ?? null
    if (hitSpr == null) {
      sprMng.activeSprite = null
      return
    }
    sprMng.activeSprite = hitSpr

    rect = hitSpr.rect.clone()
    // 保留 10px，避免移出边界，无法拖回来
    limit = {
      xl: -rect.w + 10,
      xr: cvsEl.width - 10,
      yt: -rect.h + 10,
      yb: cvsEl.height - 10
    }

    startX = clientX
    startY = clientY
    window.addEventListener('mousemove', onMouseMove)
    window.addEventListener('mouseup', () => {
      clearWindowEvt()
      if (sprMng.activeSprite != null) {
        offScale()
        offScale = scalableSprite(cvsEl, sprMng.activeSprite)
      }
    })
  }

  const onMouseMove = (evt: MouseEvent): void => {
    if (hitSpr == null || rect == null || limit == null) return
    const { clientX, clientY } = evt
    let newX = rect.x + (clientX - startX) / cvsRatio.w
    let newY = rect.y + (clientY - startY) / cvsRatio.h

    // 限制不能完全拖拽出容器边界
    newX = newX <= limit.xl ? limit.xl : newX >= limit.xr ? limit.xr : newX
    newY = newY <= limit.yt ? limit.yt : newY >= limit.yb ? limit.yb : newY
    hitSpr.rect.x = newX
    hitSpr.rect.y = newY
  }

  cvsEl.addEventListener('mousedown', onCvsMouseDown)

  const clearWindowEvt = (): void => {
    window.removeEventListener('mousemove', onMouseMove)
    window.removeEventListener('mouseup', clearWindowEvt)
  }

  return (): void => {
    clearWindowEvt()
    offAddSpr()
    offScale()
    cvsEl.removeEventListener('mousedown', onCvsMouseDown)
  }
}

/**
 * 让sprite可缩放
 */
export function scalableSprite (
  cvsEl: HTMLCanvasElement,
  s: BaseSprite
): () => void {
  const cvsRatio = {
    w: cvsEl.clientWidth / cvsEl.width,
    h: cvsEl.clientHeight / cvsEl.height
  }
  const { ctrls } = s.rect

  let startX = 0
  let startY = 0
  // 点击事件命中的控制点名字
  let ctrlKey: TCtrlKey | null = null
  let ctrlRect: Rect | null = null
  const onCvsMouseDown = (evt: MouseEvent): void => {
    // 鼠标左键才能拖拽移动
    if (evt.button !== 0) return
    const { offsetX, offsetY, clientX, clientY } = evt
    // 将鼠标点击偏移坐标映射成 canvas 坐标，然后映射成相对 sprite 的中心坐标
    // 因为 ctrls 是相对 sprite 中心点定位的
    const ofx = offsetX / cvsRatio.w - s.rect.center.x
    const ofy = offsetY / cvsRatio.h - s.rect.center.y
    const it = Object.entries(ctrls)
      .find(([, rect]) => rect.checkHit(ofx, ofy))
    if (it == null) return

    ctrlKey = it[0] as TCtrlKey
    ctrlRect = it[1]
    startX = clientX
    startY = clientY
    window.addEventListener('mousemove', onMouseMove)
    window.addEventListener('mouseup', clearWindowEvt)
  }
  cvsEl.addEventListener('mousedown', onCvsMouseDown)

  const startRect = s.rect.clone()
  const onMouseMove = (evt: MouseEvent): void => {
    if (ctrlKey == null || ctrlRect == null) return
    const { clientX, clientY } = evt
    const deltaX = (clientX - startX) / cvsRatio.w
    const deltaY = (clientY - startY) / cvsRatio.h

    // 对角线上的点是等比例缩放，key 的长度为 2
    // const hanler = ctrlKey.length === 1
    //   ? stretchScale
    //   : fixedRatioScale
    const { x, y, w, h } = startRect
    const { incW, incH, incS, rotateAngle } = stretchScale({
      deltaX,
      deltaY,
      angle: s.rect.angle,
      ctrlKey
    })

    // 最小缩放限定
    const minSize = 10
    let newW = w + incW
    newW = newW < minSize ? minSize : newW
    let newH = h + incH
    newH = newH < minSize ? minSize : newH

    const newCntX = incS / 2 * Math.cos(rotateAngle) + x + w / 2
    const newCntY = incS / 2 * Math.sin(rotateAngle) + y + h / 2

    const newX = newCntX - newW / 2
    const newY = newCntY - newH / 2

    s.rect.x = newX
    s.rect.y = newY
    s.rect.w = newW
    s.rect.h = newH
  }

  const clearWindowEvt = (): void => {
    window.removeEventListener('mousemove', onMouseMove)
    window.removeEventListener('mouseup', clearWindowEvt)
  }

  return () => {
    clearWindowEvt()
    cvsEl.removeEventListener('mousedown', onCvsMouseDown)
  }
}

/**
 * 拉伸缩放
 */
function stretchScale ({
  deltaX, deltaY, angle, ctrlKey
}: { deltaX: number, deltaY: number, angle: number, ctrlKey: TCtrlKey }): {
    incW: number
    incH: number
    incS: number
    rotateAngle: number
  } {
  // 计算矩形增加的宽度
  let incS = 0
  let incW = 0
  let incH = 0
  let rotateAngle = angle
  if (ctrlKey === 'l' || ctrlKey === 'r') {
    incS = deltaX * Math.cos(angle) + deltaY * Math.sin(angle)
    // l 缩放是反向的
    incW = incS * (ctrlKey === 'l' ? -1 : 1)
  } else if (ctrlKey === 't' || ctrlKey === 'b') {
    // 计算矩形增加的宽度，旋转坐标系让x轴与角度重合，鼠标位置在x轴的投影（x值）即为增加的高度
    rotateAngle = angle - Math.PI / 2
    incS = deltaX * Math.cos(rotateAngle) + deltaY * Math.sin(rotateAngle)
    incH = incS * (ctrlKey === 'b' ? -1 : 1)
  }

  return { incW, incH, incS, rotateAngle }
}

/**
 * 等比例缩放
 */
function fixedRatioScale ({
  ox, oy, angle, ctrlKey, rect
}: { ox: number, oy: number, angle: number, ctrlKey: TCtrlKey, rect: Rect }): {
    incW: number
    incH: number
    incS: number
    rotateAngle: number
  } {
  // 对角线角度
  const diagonalAngle = Math.atan2(rect.h, rect.w)
  // 坐标系旋转角度， lb->rt的对角线的初始角度为负数，所以需要乘以-1
  const rotateAngle = (
    ctrlKey === 'lt' || ctrlKey === 'rb' ? 1 : -1
  ) * diagonalAngle + angle
  // 旋转坐标系让x轴与对角线重合，鼠标位置在x轴的投影（x值）即为增加的长度
  const incS = ox * Math.cos(rotateAngle) + oy * Math.sin(rotateAngle)
  // lb lt 缩放值是反向
  const coefficient = ctrlKey === 'lt' || ctrlKey === 'lb' ? -1 : 1
  // 等比例缩放，增加宽高等于长度乘以对应的角度函数
  // 因为等比例缩放，中心及被拖拽的点，一定在对角线上
  const incW = incS * Math.cos(diagonalAngle) * coefficient
  const incH = incS * Math.sin(diagonalAngle) * coefficient

  return { incW, incH, incS, rotateAngle }
}
