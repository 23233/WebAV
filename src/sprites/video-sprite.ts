import { BaseSprite } from './base-sprite'

interface IVideoSpriteOpts {
  audioCtx?: AudioContext
}

export class VideoSprite extends BaseSprite {
  #videoEl: HTMLVideoElement | null = null

  constructor (name: string, source: MediaStream, opts: IVideoSpriteOpts = {}) {
    super(name)
    this.initReady = this.#init(source, opts)
  }

  async #init (ms: MediaStream, opts: IVideoSpriteOpts): Promise<void> {
    const audioMS = new MediaStream()
    ms.getAudioTracks().forEach((track) => {
      // 给视频消音
      ms.removeTrack(track)
      audioMS.addTrack(track)
    })

    this.#videoEl = await mediaStream2Video(ms)
    await this.#videoEl.play()

    this.rect.w = this.#videoEl.videoWidth
    this.rect.h = this.#videoEl.videoHeight

    if (opts.audioCtx != null && audioMS.getAudioTracks().length > 0) {
      const audioSource = opts.audioCtx.createMediaStreamSource(audioMS)
      this.audioNode = opts.audioCtx.createGain()
      audioSource.connect(this.audioNode)
    }
  }

  render (ctx: CanvasRenderingContext2D): void {
    if (this.#videoEl == null) return
    super.render(ctx)
    const { w, h } = this.rect
    ctx.drawImage(this.#videoEl, -w / 2, -h / 2, w, h)
    ctx.resetTransform()
  }

  get volume (): number {
    return this.audioNode?.gain.value ?? 0
  }

  set volume (v: number) {
    if (this.audioNode == null) return
    this.audioNode.gain.value = v
  }

  destory (): void {
    this.#videoEl?.remove()
    this.#videoEl = null
  }
}

async function mediaStream2Video (
  stream: MediaStream
): Promise<HTMLVideoElement> {
  const video = document.createElement('video')

  let timer: number

  video.srcObject = stream

  return await new Promise((resolve, reject) => {
    let failed = false
    video.addEventListener('loadeddata', () => {
      if (failed) return
      clearTimeout(timer)
      resolve(video)
    })
    timer = window.setTimeout(() => {
      failed = true
      reject(new Error('video load failed'))
    }, 2000)
  })
}
