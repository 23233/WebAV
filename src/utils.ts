export async function mediaStream2Video (
  stream: MediaStream
): Promise<HTMLVideoElement> {
  const video = document.createElement('video')

  let timer: number

  video.srcObject = stream

  return await new Promise((resolve, reject) => {
    let failed = false
    video.addEventListener('loadedmetadata', () => {
      if (failed) return
      clearTimeout(timer)
      video.play()
        .then(() => resolve(video))
        .catch(reject)
    })
    timer = window.setTimeout(() => {
      failed = true
      reject(new Error('video load failed'))
    }, 2000)
  })
}

export function createEl (tagName: string): HTMLElement {
  return document.createElement(tagName)
}
