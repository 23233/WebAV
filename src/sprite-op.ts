import { BaseSprite } from './sprites'

export function draggabelSprite (
  cvsEl: HTMLCanvasElement,
  spriteList: BaseSprite[]
): () => void {
  const onMouseMove = (): void => {}

  const ratio = {
    w: cvsEl.clientWidth / cvsEl.width,
    h: cvsEl.clientHeight / cvsEl.height
  }
  // 排在后面的层级更高
  const sprList = spriteList.reverse()
  let startX = 0
  let startY = 0
  // 按下cvs，寻找选中的 sprite，监听移动事件
  const onCvsMouseDown = (evt: MouseEvent): void => {
    window.addEventListener('mousemove', onMouseMove)
    window.addEventListener('mouseup', clearWindowEvt)

    // 鼠标左键才能拖拽移动
    if (evt.button !== 0) return
    const { offsetX, offsetY, clientX, clientY } = evt
    const hitSpr = sprList.find(s => s.checkHit(
      offsetX / ratio.w,
      offsetY / ratio.h
    ))
    // rect = getHitRect(offsetX, offsetY)
    if (hitSpr == null) return
    // todo: 保留 10px，避免移出边界，无法拖回来
    // limit = {
    //   xl: -rect.w + 10,
    //   xr: view.width - 10,
    //   yt: -rect.h + 10,
    //   yb: view.height - 10
    // }

    startX = clientX
    startY = clientY
  }

  cvsEl.addEventListener('mousedown', onCvsMouseDown)

  const clearWindowEvt = (): void => {
    window.removeEventListener('mousemove', onMouseMove)
    window.removeEventListener('mouseup', clearWindowEvt)
  }

  return (): void => {
    clearWindowEvt()
    cvsEl.removeEventListener('mousedown', onCvsMouseDown)
  }
}
