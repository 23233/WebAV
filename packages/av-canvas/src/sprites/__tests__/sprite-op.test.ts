import { beforeEach, describe, expect, test, vi } from 'vitest'
import { crtMSEvt4Offset } from '../../__tests__/mock'
import { SpriteManager } from '../sprite-manager'
import { createEl } from '../../utils'
import { draggabelSprite } from '../sprite-op'
import { VideoSprite } from '../video-sprite'

let cvsEl = createEl('canvas') as HTMLCanvasElement
const cvsRatio = {
  w: 1,
  h: 1
}
let sprMng = new SpriteManager()
beforeEach(() => {
  sprMng = new SpriteManager()
  cvsEl = createEl('canvas') as HTMLCanvasElement
  vi.spyOn(cvsEl, 'clientWidth', 'get').mockImplementation(() => 900)
  vi.spyOn(cvsEl, 'clientHeight', 'get').mockImplementation(() => 500)
  cvsEl.width = 1920
  cvsEl.height = 1080
  cvsRatio.w = 900 / 1920
  cvsRatio.h = 500 / 1080
})

describe('draggabelSprite', () => {
  test('canvas on mousedown', () => {
    const spyAEL = vi.spyOn(cvsEl, 'addEventListener')
    const spyREL = vi.spyOn(cvsEl, 'removeEventListener')

    const clear = draggabelSprite(cvsEl, sprMng)
    expect(spyAEL).toBeCalledWith('mousedown', expect.any(Function))
    expect(clear).toBeInstanceOf(Function)

    clear()
    expect(spyREL).toBeCalledWith('mousedown', expect.any(Function))
  })

  test('window on mouse event', async () => {
    const spyAEL = vi.spyOn(window, 'addEventListener')
    const spyREL = vi.spyOn(window, 'removeEventListener')
    const vs = new VideoSprite('vs', new MediaStream())
    vi.spyOn(vs.rect, 'checkHit').mockReturnValue(true)
    await sprMng.addSprite(vs)
    sprMng.activeSprite = vs
    const clear = draggabelSprite(cvsEl, sprMng)
    cvsEl.dispatchEvent(new MouseEvent('mousedown'))

    expect(spyAEL).toBeCalledTimes(2)
    expect(spyAEL).toHaveBeenNthCalledWith(1, 'mousemove', expect.any(Function))
    expect(spyAEL).toHaveBeenNthCalledWith(2, 'mouseup', expect.any(Function))

    clear()
    expect(spyREL).toHaveBeenNthCalledWith(1, 'mousemove', expect.any(Function))
    expect(spyREL).toHaveBeenNthCalledWith(2, 'mouseup', expect.any(Function))
  })

  test('move sprite', async () => {
    const vs = new VideoSprite('vs', new MediaStream())
    vs.rect.w = 100
    vs.rect.h = 100

    await sprMng.addSprite(vs)
    sprMng.activeSprite = vs

    const clear = draggabelSprite(cvsEl, sprMng)
    cvsEl.dispatchEvent(crtMSEvt4Offset('mousedown', 50, 50))

    window.dispatchEvent(new MouseEvent('mousemove', {
      clientX: 100,
      clientY: 100
    }))
    expect(vs.rect.x).toBe(100 / cvsRatio.w)
    expect(vs.rect.y).toBe(100 / cvsRatio.h)

    // 鼠标移动超出边界
    window.dispatchEvent(new MouseEvent('mousemove', {
      clientX: 10000,
      clientY: 10000
    }))
    // sprite 至少保留10px在可视区域内
    expect(vs.rect.x).toBe(cvsEl.width - 10)
    expect(vs.rect.y).toBe(cvsEl.height - 10)

    clear()
  })
})

describe('scale sprite', () => {
  test('drag right ctrl', async () => {
    const vs = new VideoSprite('vs', new MediaStream())
    await sprMng.addSprite(vs)
    sprMng.activeSprite = vs
    vs.rect.w = 100
    vs.rect.h = 100

    // 激活 sprite
    const clear = draggabelSprite(cvsEl, sprMng)
    cvsEl.dispatchEvent(crtMSEvt4Offset('mousedown', 0, 0))
    expect(sprMng.activeSprite).toBe(vs)

    window.dispatchEvent(new MouseEvent('mouseup'))
    // 命中 right ctrl
    cvsEl.dispatchEvent(crtMSEvt4Offset(
      'mousedown',
      100 * cvsRatio.w,
      50 * cvsRatio.h
    ))
    window.dispatchEvent(new MouseEvent('mousemove', {
      clientX: 100,
      clientY: 100
    }))
    // 拖拽 right ctrl 缩放 rect 的宽度
    expect(vs.rect.w).toBe(100 + 100 / cvsRatio.w)

    clear()
  })

  test('drag rb(bottom right) ctrl', async () => {
    const vs = new VideoSprite('vs', new MediaStream())
    await sprMng.addSprite(vs)
    sprMng.activeSprite = vs
    vs.rect.w = 100
    vs.rect.h = 100

    // 激活 sprite
    const clear = draggabelSprite(cvsEl, sprMng)
    cvsEl.dispatchEvent(crtMSEvt4Offset('mousedown', 0, 0))
    expect(sprMng.activeSprite).toBe(vs)

    window.dispatchEvent(new MouseEvent('mouseup'))
    // 命中 right ctrl
    cvsEl.dispatchEvent(crtMSEvt4Offset(
      'mousedown',
      100 * cvsRatio.w,
      100 * cvsRatio.h
    ))
    window.dispatchEvent(new MouseEvent('mousemove', {
      clientX: 100,
      clientY: 100
    }))
    expect(vs.rect).toMatchSnapshot()

    clear()
  })
})

describe('rotate sprite', () => {
  test('rotate sprite', async () => {
    const vs = new VideoSprite('vs', new MediaStream())
    await sprMng.addSprite(vs)
    sprMng.activeSprite = vs
    vs.rect.w = 100
    vs.rect.h = 100

    // 激活 sprite
    const clear = draggabelSprite(cvsEl, sprMng)
    cvsEl.dispatchEvent(crtMSEvt4Offset('mousedown', 0, 0))
    expect(sprMng.activeSprite).toBe(vs)

    window.dispatchEvent(new MouseEvent('mouseup'))
    // 命中 rotate ctrl
    const { center, ctrls: { rotate } } = vs.rect
    cvsEl.dispatchEvent(crtMSEvt4Offset(
      'mousedown',
      (rotate.x + center.x) * cvsRatio.w,
      (rotate.y + center.y) * cvsRatio.h
    ))
    window.dispatchEvent(new MouseEvent('mousemove', {
      clientX: 100,
      clientY: 100
    }))
    expect(vs.rect.angle).toMatchSnapshot()

    window.dispatchEvent(new MouseEvent('mousemove', {
      clientX: 100,
      clientY: 200
    }))
    expect(vs.rect.angle).toMatchSnapshot()

    clear()
  })
})
