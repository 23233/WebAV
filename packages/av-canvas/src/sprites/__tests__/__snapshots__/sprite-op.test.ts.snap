// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rotate sprite > rotate sprite 1`] = `2.3580805686801525`;

exports[`rotate sprite > rotate sprite 2`] = `2.7330338359124564`;

exports[`scale sprite > drag rb(bottom right) ctrl 1`] = `
G {
  "angle": 0,
  "h": 314.66666666666663,
  "master": null,
  "w": 314.6666666666667,
  "x": 0,
  "y": 0,
}
`;
